### DISCUSSION TITLE
Job for auto callbook lookup for QSOs

### DISCUSSION THREAD

**Discussion by filipmelik** [wavelog/wavelog#802] [Ideas]
FLE is nice way how to log contacts fast. But when you save the QSO, there is just the basic information. It would be nice to have the rest of the fields populated from callbook. 

Looking through the source code, I have found the [unfinished feature of "Fetch from callbook" option](https://github.com/wavelog/wavelog/blob/646ccf2783042ba6390cea1ecc5ab4d3996b7b3c/assets/js/sections/simplefle.js#L187) in the FLE's "Options" menu. However, thinking about this, I am not quite sure how it should be implemented regarding UX and "stability".  I think there are following approaches how to tackle this (maybe there is more):

1. When this future option is turned on, it could fetch the data upon pressing the enter and store in some browser localstorage cache or something. And when the logging session ended it will save the fields fetched from callbook along with the current fields. However, this does not seem right, since you can hit "Reload QSO list" button or similar, which would trigger requesting callbook everytime you do this.
2. When the logging session is ended, and you hit the "Save to Wavelog" button, it could fetch the data, since you know the list of QSO's you have entered is final and will not change. However, this approach does not seem right to me. In case of many QSO's, It would significantly delay the saving to my log, apart from the fact when you are on slow internet, it could potentially screw things up all together and losing the log if something goes south.
3. We could ditch the currently unfinished option from FLE settings alltogether and make a similar functionality: Completely new cronjob that will do this. But not just for the FLE QSOs, but globally for all the QSO's in my log. When the field in my log will be populated, nothing will be done (not overwiting the existing data). And subsequently, if for example name and QTH fields are missing (probably because these are QSOs added by FLE ), those fields will be populated from callbooks. I like this approach the most and I am ready to tackle this in pull request.

Side note: I Know there is option to update data from callbook manually in the "Advanced logbook", but it seems clunky to do this, since one of the killer features of wavelog (and web application in general), is ability to have automatic jobs to do stuff (read "executing cron jobs").

WDYT?

**Comment by HB9HIL** [comment]
hey @filipmelik 

I thought about something like you mentioned in the third option aswell. My issue here, and still not sure how to solve this, is the generated load on the system due to the Callbook Loads. Imagine a instance with 10-20 users and more then 100'000 QSOs. This is generating an extreme load which is not necessary. 

It is the way more efficient way to trigger the callbook lookups where it's needed. So for SimpleFLE during save or maybe in the future for API incoming QSOs.

**Reply by filipmelik** [reply]
Yeah, fair point. I thought about this as well. I am not sure what are all the cases that cause the QSO to have only basic information, I know only about:
1. SFLE input
2. logging through the search field (if turned on in settings)

Obviously what could be done is something like flagging QSOs made this way in DB and this potential sync cron would download only data for those marked QSOs and then "unflagging" them when done. WDYT? 

You mentioned 
> or maybe in the future for API incoming QSOs

You think that this could be done directly using the "QSO save API" in the moment of saving? Can you elaborate pls? Because if this is what you are thinking about, it could be combined with my idea above - the "QSO save API" would just flag the QSOs (so it does not block the saving until the callbook data is fetched) and the rest will be done with to cronjob.

**Reply by HB9HIL** [reply]
I don't see a cronjob for this for several reasons

1. The already mentioned load. Even if you only handle flagged QSO's you will need to get those flagged QSO's first. Sure this is not a higher load then editing them but still you need to iterate through the complete database. 
2. We have several thirdparty services included which are synced by cronjobs. So if my LoTW cronjob is syncing at 3AM in the morning, I would have to make sure, that the "Missing QSO data" cron is running BEFORE that. So the already updated QSO's get uploaded. Same for QRZ.com and the others. 

The main goal is to "automate" the process of fetching missing data from the callbook. This does not mean, that you need a cronjob for that. 

There are the following ways of QSO's into Wavelog

- QSO logging in webui 
- Contest logging in webui
- SimpleFLE in webui
- QSO API 
- ADIF Import

The "only thing" we need to do is to create a switch which performs a callbook lookup when new data is coming in. This switch can be either in user_options (see user_options_model) or globally in global options. I would prefer the first one. 

@int2001 and I already thought about some kind of backend worker which performs jobs Independent of the CI session. It was an idea in the past to handle the ADIF import with that. Something like upload the file and parse it in the background without blocking the WebUI. Unfortunately the nature of PHP does not allow a daemon-like service so the only way to achive this is so _simulate_ this with a background worker (triggered by the master cron every minute?). This worker could also then perform tasks like the callbook lookup for missing data. A lot of ideas and possible ways.. However we try to not refactor everything from scratch. We are only four guys which do this as a hobby 😄 Step by Step; Small PR's and Easy to review changes

**Comment by HB9HIL** [comment]
In addition to that there is a big process which needs to be done first, before we can add new cronjobs (#339).
